### PR TITLE
feat(v1 pool): adding support for v1 pool grpc

### PR DESCRIPTION
**Pull request by abhilashshetty04**
Changes in this PR allows Control plane to use v1 apis for Pool operations.

### CODE REVIEW

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/core/src/controller/wrapper.rs [hunk 1]
```diff
@@ -1225,6 +1247,13 @@ pub(crate) fn rpc_pool_to_agent(rpc_pool: &rpc::io_engine::Pool, id: &NodeId) ->
     pool
 }
 
+/// convert rpc pool to a agent pool
+pub(crate) fn v1_rpc_pool_to_agent(rpc_pool: &rpc::v1::pool::Pool, id: &NodeId) -> PoolState {
+    let mut pool = rpc_pool.to_agent();
+    pool.node = id.clone();
+    pool
+}
+
```
Comment (Ab-hishek) [inline control-plane/agents/core/src/controller/wrapper.rs]: Can we shift this to `msg_translation/v1.rs` file. Better to keep any conversion related to versions in their respective files.

File: control-plane/agents/common/src/msg_translation/v1.rs [hunk 2]
```diff
@@ -334,3 +334,42 @@ pub fn rpc_nexus_to_child_agent(
...
+
+impl AgentToIoEngine for transport::CreatePool {
+    type IoEngineMessage = v1_rpc::pool::CreatePoolRequest;
+    ///This converts Control plane CreatePool struct to IO Engine gRPC message.
+    fn to_rpc(&self) -> Self::IoEngineMessage {
+        Self::IoEngineMessage {
+            name: self.id.clone().into(),
+            disks: self.disks.iter().map(|d| d.to_string()).collect(),
+            uuid: None,
+            pooltype: 0,
```
Comment (tiagolobocastro) [inline control-plane/agents/common/src/msg_translation/v1.rs]: Please use the enum v1_rpc::abcd as i32.

File: rpc/src/lib.rs [hunk 3]
```diff
@@ -137,4 +137,12 @@ pub mod v1 {
             RemoveChildNexusResponse, UnpublishNexusRequest, UnpublishNexusResponse,
         };
     }
+
+    /// V1 Pool autogenerated grpc code
```
Comment (tiagolobocastro) [inline rpc/src/lib.rs]: end comment with period `.`

File: control-plane/agents/core/src/controller/wrapper.rs [hunk 4]
```diff
@@ -1420,6 +1439,13 @@ pub(crate) fn rpc_pool_to_agent(rpc_pool: &rpc::io_engine::Pool, id: &NodeId) ->
     pool
 }
 
+/// convert rpc pool to a agent pool
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/wrapper.rs]:
```suggestion
/// Convert rpc pool to an agent pool.
```

File: control-plane/agents/core/src/controller/grpc.rs [hunk 5]
```diff
@@ -115,11 +118,15 @@ pub(crate) type ReplicaClient = rpc::v1::replica::replica_rpc_client::ReplicaRpc
 /// V1 NexusClient
 pub(crate) type NexusClient = rpc::v1::nexus::nexus_rpc_client::NexusRpcClient<Channel>;
 
+///V1 PoolClient
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/grpc.rs]:
```suggestion
/// The V1 PoolClient.
```

File: control-plane/agents/core/src/controller/grpc.rs [hunk 6]
```diff
@@ -115,11 +118,15 @@ pub(crate) type ReplicaClient = rpc::v1::replica::replica_rpc_client::ReplicaRpc
...
+pub(crate) type PoolCLient = rpc::v1::pool::pool_rpc_client::PoolRpcClient<Channel>;
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/grpc.rs]: PoolClient

File: control-plane/agents/common/src/msg_translation/v1.rs [hunk 7]
```diff
@@ -334,3 +334,42 @@ pub fn rpc_nexus_to_child_agent(
...
+        }
+    }
+}
+
+impl AgentToIoEngine for transport::DestroyPool {
+    type IoEngineMessage = v1_rpc::pool::DestroyPoolRequest;
+    ///This converts Control plane DeletePool struct to IO Engine gRPC message.
```
Comment (tiagolobocastro) [inline control-plane/agents/common/src/msg_translation/v1.rs]: here and above, please add space between /// and first letter.

**Review by tiagolobocastro — COMMENTED** [review]
Please squash your commits before the merge.

File: rpc/src/lib.rs (hunk 3 above)
Comment (abhilashshetty04) [reply rpc/src/lib.rs]: Incorporated

File: control-plane/agents/core/src/controller/grpc.rs (hunk 5 above)
Comment (abhilashshetty04) [reply control-plane/agents/core/src/controller/grpc.rs]: Incorporated

File: control-plane/agents/common/src/msg_translation/v1.rs (hunk 7 above)
Comment (abhilashshetty04) [reply control-plane/agents/common/src/msg_translation/v1.rs]: Incorporated

File: control-plane/agents/core/src/controller/grpc.rs (hunk 6 above)
Comment (abhilashshetty04) [reply control-plane/agents/core/src/controller/grpc.rs]: Incorporated

File: control-plane/agents/core/src/controller/wrapper.rs (hunk 1 above)
Comment (abhilashshetty04) [reply control-plane/agents/core/src/controller/wrapper.rs]: Incorporated

File: control-plane/agents/core/src/controller/wrapper.rs (hunk 4 above)
Comment (abhilashshetty04) [reply control-plane/agents/core/src/controller/wrapper.rs]: Incorporated

File: control-plane/agents/common/src/msg_translation/v1.rs (hunk 2 above)
Comment (abhilashshetty04) [reply control-plane/agents/common/src/msg_translation/v1.rs]: Incorporated

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/wrapper.rs [hunk 8]
```diff
@@ -833,13 +834,18 @@ impl ClientOps for Arc<tokio::sync::RwLock<NodeWrapper>> {
...
+        let destroy_response = dataplane.destroy_pool(request).await;
         let mut ctx = dataplane.reconnect(GETS_TIMEOUT).await?;
         self.update_pool_states(ctx.deref_mut()).await?;
-        Ok(())
+        match destroy_response {
+            Ok(_) => Ok(()),
+            Err(_) => Err(SvcError::PoolNotFound {
+                pool_id: request.id.clone(),
+            }),
+        }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/wrapper.rs]: I'm confused by this, seems all errors become a PoolNotFound error?

File: control-plane/agents/src/bin/core/controller/wrapper.rs [hunk 9]
```diff
@@ -1107,7 +1123,16 @@ impl ClientOps for GrpcClientLocked {
...
-                unimplemented!()
+                let _ = self
+                    .client_v1()?
+                    .pool_client()
+                    .destroy_pool(v1_conversion::to_rpc(request))
+                    .await
+                    .context(GrpcRequestError {
+                        resource: ResourceKind::Pool,
+                        request: "destroy_pool",
+                    })?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/wrapper.rs:1173]: don't we need to handle error not found here?

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/grpc.rs [hunk 10]
```diff
@@ -131,6 +136,10 @@ impl MayaClientV1 {
     pub(crate) fn nexus(self) -> NexusClient {
         self.nexus
     }
+    /// Get the v1 pool client.
+    pub(crate) fn pool_client(&self) -> PoolClient {
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/controller/grpc.rs]: Can we change this to `pool` to maintain the similar nomeclature.

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 8 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: Yes, At this time dataplane returns error in only 2 scenarios. 1. When uuid provided in DestroyPoolRequest doesnt match with uuid of the object in dataplane. 2. When Pool is not found
1st scenario is unlikely as we do not create Pools with uuid. DestroyPoolRequest is also sent with None for uuid.
However, It doesnt seem right to map any error to one type. So i will add logic in Control plane to check if pool exist and return error.

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: I don't think we should make such assumptions, eg: the error may be a transport error or a timeout and in such case it's really confusing if we report a pool not found error.
I think we should simply return the error we find, and specifically map NotFound to Ok(()), for now. Later I think we probably want to handle notFound differently but for now we should map to Ok(()) to be compatible with the current code.

Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: Will look into modifying errors returned by dataplane.

Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs]: This PR https://github.com/openebs/mayastor/pull/1192/files will make sure dataplane returns more specific errors. Incorporated changes you requested here also.

File: control-plane/agents/src/bin/core/controller/grpc.rs (hunk 10 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/grpc.rs]: Incorporated

File: control-plane/agents/src/bin/core/controller/wrapper.rs (hunk 9 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/wrapper.rs:1173]: This is just a conversion from gRPC tonic status to Service errors. We are matching errors once we have Service error.

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/src/bin/core/pool/pool_operations.rs [hunk 11]
```diff
@@ -30,6 +30,11 @@ impl ResourceLifecycle for OperationGuardArc<PoolSpec> {
                 node_id: request.node.to_string(),
             });
         }
+        if request.disks.len() != 1 {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/pool/pool_operations.rs:33]: Do we need this check in the control plane?

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/pool/pool_operations.rs:33]: Ok I think it's fine to have it for now, though we might need to change this if we ever have different pool types.

**Review by tiagolobocastro — APPROVED** [review]

**Review by vharsh — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**